elifepubmed/generate.py

```diff
@@ -601,6 +601,45 @@ def set_grants(parent, poa_article):
                 set_object(parent, "grant", params)
 
 
+def dataset_assigning_authority(assigning_authority, uri):
+    """precise assigning_authority value considering the uri in some cases"""
+    authority_map = OrderedDict([('NCBI', [])])
+    # tuple in form: uri hint, new assigning authority
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/geo', 'NCBI:geo'))
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/projects/gap', 'NCBI:dbgap'))
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/nuccore', 'NCBI:nucleotide'))
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/sra', 'NCBI:sra'))
+
+    if assigning_authority in authority_map:
+        for hint, new_value in authority_map.get(assigning_authority):
+            if hint in uri:
+                assigning_authority = new_value
+    return assigning_authority
+
+
+def dataset_details(dataset):
+    """
+    return assigning_authority and id value for dataset
+    consider the uri value, it may change the assigning_authority
+
+    :param dataset: Dataset object
+    :returns: string assigning authority of the dataset, string id is the uri or doi
+    """
+    assigning_authority = dataset_assigning_authority(dataset.assigning_authority, dataset.uri)
+    id_value = utils.first_property(dataset, ["doi", "accession_id"])
```

Comment:
Is this the only use of `first_property`? Would `dataset.doi or dataset.accession_id` perhaps be slightly more clear?

Reply:
@de-code you're on to something I like here. `first_property()` / `first_attribute()` was a first go at it, and was only used in this place, and I can see how to refactor for a cleaner expression of this.

Reply:
I was going to use a list comprehension, then I ended up just using `firstnn()` we've used for years from `elifetools.utils`, it is already tested and working for finding a first non-None value.

Reply:
Personally I wouldn't know what `firstnn` means without reading the documentation. Neither would `first_attribute`. With just two fields and one use-case, just writing it out as `dataset.doi or dataset.accession_id` or `dataset.doi if dataset.doi else dataset.accession_id` would be more readable to me. But maybe as a regular elife bot user, `firstnn` is equally readable.